File: API/CC_API/src/Databases/Table.c
#include "Databases/Table.h"

WebFrameworkException executeQuery(Table implementation, const char* query, const SQLValue* values, size_t size, SQLResult* result)
{
	WebFrameworkException exception = NULL;

	typedef void* (*executeQuery)(void* implementation, const char* query, const void** values, size_t size, void** exception);

	*result = CALL_CLASS_MEMBER_WEB_FRAMEWORK_FUNCTION(executeQuery, query, (const void**)values, size, &exception);

	return exception;
}

WebFrameworkException deleteSQLResult(Table implementation, SQLResult result)
{
	WebFrameworkException exception = NULL;

	typedef void (*deleteSQLResult)(void* implementation, void* result, void** exception);

	CALL_CLASS_MEMBER_WEB_FRAMEWORK_FUNCTION(deleteSQLResult, result, &exception);

	return exception;
}

WebFrameworkException getTableName(Table implementation, const char** result)
{
	WebFrameworkException exception = NULL;

	typedef const char* (*getTableName)(void* implementation, void** exception);

	*result = CALL_CLASS_MEMBER_WEB_FRAMEWORK_FUNCTION(getTableName, &exception);

	return exception;
}

File: API/CSharp_API/Databases/Table.cs
namespace Framework;

using Framework.Exceptions;
using Framework.Utility;
using System.Runtime.InteropServices;

public sealed partial class Table(IntPtr implementation)
{
	private readonly IntPtr implementation = implementation;

	[LibraryImport(DLLHandler.LIBRARY_NAME, StringMarshalling = StringMarshalling.Utf8)]
	private static partial IntPtr executeQuery(IntPtr implementation, string query, [In] IntPtr[]? values, nuint size, ref IntPtr exception);

	[LibraryImport(DLLHandler.LIBRARY_NAME)]
	private static partial IntPtr getTableName(IntPtr implementation, ref IntPtr exception);

	[LibraryImport(DLLHandler.LIBRARY_NAME)]
	private static partial void deleteSQLResult(IntPtr tableImplementation, IntPtr implementation, ref IntPtr exception);

	public SqlResult ExecuteQuery(string query, IList<SqlValue>? values = null)
	{
		IntPtr exception = IntPtr.Zero;
		IntPtr result;
		
		if (values != null)
		{
			IntPtr[] pointers = new IntPtr[values.Count];

			for (int i = 0; i < values.Count; i++)
			{
				pointers[i] = values[i].implementation;
			}

			result = executeQuery(implementation, query, pointers, (nuint)pointers.Length, ref exception);
		}
		else
		{
			result = executeQuery(implementation, query, null, 0, ref exception);
		}

		if (exception != IntPtr.Zero)
		{
			throw new WebFrameworkException(exception);
		}

		SqlResult sqlResult = new(result);

		deleteSQLResult(implementation, result, ref exception);

		if (exception != IntPtr.Zero)
		{
			throw new WebFrameworkException(exception);
		}

		return sqlResult;
	}

	public string GetTableName()
	{
		IntPtr exception = IntPtr.Zero;
		string result = Marshal.PtrToStringUTF8(getTableName(implementation, ref exception))!;

		if (exception != IntPtr.Zero)
		{
			throw new WebFrameworkException(exception);
		}

		return result;
	}
}
